### PR TITLE
Allow and implement recursive static variables.

### PR DESCRIPTION
**Pull request by eefriedman**
**_Edit: Fixed now.**_ I'm pretty sure the way I'm using LLVMReplaceAllUsesWith here is
unsafe... but before I figure out how to fix that, I'd like a
reality-check: is this actually useful?

### CODE REVIEW

**Comment by eddyb** [comment]
This would be really neat for `&'static` cycles, e.g. a safe doubly linked static list.

**Comment by eefriedman** [comment]
Updated; should be safe now.  Also added an extra commit to remove the restriction on types which can't be instantiated, to allow implementing a static doubly-linked list.

**Comment by alexcrichton** [comment]
cc @rust-lang/lang, not sure if this is a large enough change to warrant an RFC, it'd probably be nice to have a "static specification" but that may be a bit of a pipe dream.

**Comment by nrc** [comment]
I don't think it needs an RFC and I think it is a good idea. I do think it probably deserves some eyes on it, so I created an announcement in our internals discourse: https://internals.rust-lang.org/t/heads-up-allowing-recursive-static-variables/2309

**Comment by P1start** [comment]
Are nonsensical cycles like `static FOO: T = FOO;` and `static FOO: T = BAR; static BAR: T = FOO;` still disallowed by this PR?

Also, perhaps the tests that are deleted in this PR should instead just be moved to `run-pass`.

**Comment by eefriedman** [comment]
@P1start The initializer for a static item can't read the contents of any static item, so `static FOO: T = FOO;` is still rejected.

Not sure the tests are particularly useful, but I guess I could move them.

**Comment by eefriedman** [comment]
Tests re-added.

**Comment by pnkfelix** [comment]
Surely this should at least be feature gated?

**Comment by pnkfelix** [comment]
(that is, I can understand side-stepping the RFC process, though I'm not 100% sure I agree with doing so, even for a feature like this.  But with or without an RFC, it still seems prudent to feature-gate new functionality in the language, just so that we can kick the tires on it in nightly without worry about whether it could inadvertantly leak out into the other distribution channels.)

**Comment by eefriedman** [comment]
`static_recursion` feature gate added.  I strongly doubt there will be any issues in practice, but better to be on the safe side, I guess.

**Comment by eddyb** [comment]
AFAICT you had to revert the removal of `is_instantiable`. I guess it will go away once the feature gate is lifted.

**Comment by pnkfelix** [comment]
@eefriedman sorry I have not reviewed this yet.  I hope I will get to it this week but I am not sure if I will.

**Comment by eefriedman** [comment]
@pnkfelix Ping.

**Comment by pnkfelix** [comment]
@eefriedman sorry for the delay

File: src/librustc_trans/trans/context.rs [hunk 1]
```diff
@@ -131,6 +128,11 @@ pub struct LocalCrateContext<'tcx> {
     /// Cache of closure wrappers for bare fn's.
     closure_bare_wrapper_cache: RefCell<FnvHashMap<ValueRef, ValueRef>>,
 
+    /// List of globals for static variables which need to be RAUW'ed when
```
Comment (pnkfelix) [inline src/librustc_trans/trans/context.rs]: nit: as far as I can tell the acronym "RAUW" does not occur elsewhere in our code.

So, please expand this occurrence and then include the acronym afterwards in parentheses, effectively defining the acronym.  Like so: "static variables which need to be passed through the LLVM Replace All Uses With (RAUW) functionality"

**Comment by pnkfelix** [comment]
okay looks good as far as my knowledge of LLVM goes.

r=me after next rebase (and hopefully the noted nit is addressed).

**Comment by eefriedman** [comment]
@pnkfelix Rebased.

**Pull request merged**